### PR TITLE
Fixed using subFieled Repetition to generate requested parquet sc…

### LINKED ISSUE #9248: The requested schema is not compatible with the file schema. incompatible types
SQL
`select extinfo.qris_info.merchant_account_information from extinfo_table;`
The following error occurred after upgrading from 332 to 356.
```
The requested schema is not compatible with the file schema. incompatible types: required group qris_info {
  required binary merchant_account_information (STRING);
} != optional group qris_info {
  required binary point_of_initiation_method (STRING);
  required binary merchant_account_information (STRING);
  required binary merchant_category_code (STRING);
  required binary transaction_currency (STRING);
  required binary transaction_amount (STRING);
  required binary convenience_indicator (STRING);
  required binary value_of_convenience_fee_fixed (STRING);
  required binary value_of_convenience_fee_percentage (STRING);
  required binary country_code (STRING);
  required binary merchant_name (STRING);
  required binary merchant_city (STRING);
  required binary postal_code (STRING);
  optional group additional_data_field_template {
    required binary bill_number (STRING);
    required binary mobile_number (STRING);
    required binary store_label (STRING);
    required binary loyalty_number (STRING);
    required binary reference_label (STRING);
    required binary customer_label (STRING);
    required binary terminal_label (STRING);
    required binary purpose_of_transaction (STRING);
    required binary additional_consumer_data_request (STRING);
    required group rfu_for_emvco (LIST) {
      repeated group list {
        required binary element (STRING);
      }
    }
    required group payment_system_specific_templates (LIST) {
      repeated group list {
        required binary element (STRING);
      }
    }
    required binary security (STRING);
  }
  required binary additional_data_field_raw (STRING);
  optional group merchant_information_language_template {
    required binary language_preference (STRING);
    required binary merchant_name (STRING);
    required binary merchant_city (STRING);
  }
  required binary payload_format_indicator (STRING);
}```

### PR DESCRIPTION
**Pull request by lxqfy**
Fixed #9248

### CODE REVIEW

**Review by findepi — COMMENTED** [review]
Would it be possible to have a test?

**Comment by lxqfy** [comment]
> Would it be possible to have a test?

Hi I have added a test case for this scenario. This issue only happens when the subField has a different repetition.

**Comment by lxqfy** [comment]
> Thank you for your pull request and welcome to our community. We require contributors to sign our [Contributor License Agreement](https://github.com/trinodb/cla/raw/master/Trino%20Foundation%20Individual%20CLA.pdf), and we don't seem to have you on file. In order for us to review and merge your code, please submit the signed CLA to [cla@trino.io](mailto:cla@trino.io). For more information, see https://github.com/trinodb/cla.

I have the send the CLA on 15/09/2021

**Comment by findepi** [comment]
@lxqfy the change looks correct. Wonder if this is reproducible on data produced by Hive or Spark, or is it reproducible only for "manually" created Parquet files?
FWIW i tried to reproduce this in https://github.com/trinodb/trino/pull/9286, but did not.

cc @martint on CLA.

**Comment by lxqfy** [comment]
> @lxqfy the change looks correct. Wonder if this is reproducible on data produced by Hive or Spark, or is it reproducible only for "manually" created Parquet files?
> FWIW i tried to reproduce this in #9286, but did not.
> 
> cc @martint on CLA.

@findepi the parquet files are produced with spark by another team.

**Comment by lxqfy** [comment]
Any suggestions on how should I proceed?

**Comment by findepi** [comment]
@cla-bot check

**Comment by findepi** [comment]
>  the parquet files are produced with spark by another team.

can we reproduce the problem somehow?
maybe in TestIcebergSparkCompatibility or TestHiveSparkCompatibility test?

> Any suggestions on how should I proceed?

seems like CLA bot is not happy yet.

**Comment by lxqfy** [comment]
> > the parquet files are produced with spark by another team.
> 
> can we reproduce the problem somehow?
> maybe in TestIcebergSparkCompatibility or TestHiveSparkCompatibility test?
> 
> > Any suggestions on how should I proceed?
> 
> seems like CLA bot is not happy yet.
I am not able to construct such a column with Spark-SQL as all the StructField is OPTIONAL.

File: plugin/trino-hive/src/test/java/io/trino/plugin/hive/parquet/TestParquetPageSourceFactory.java [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+                        ImmutableList.of(1, 1),
+                        ImmutableList.of("optional_level2", "required_level3"),
+                        toHiveType(INTEGER),
+                        INTEGER)), REGULAR, Optional.empty());
+        MessageType fileSchema = new MessageType("hive_schema",
+                    new GroupType(OPTIONAL, "optional_level1",
+                            new GroupType(OPTIONAL, "optional_level2",
+                                    new PrimitiveType(REQUIRED, INT32, "required_level3"))));
+        assertEquals(ParquetPageSourceFactory.getColumnType(columnHandle, fileSchema, true).get(), fileSchema.getType("optional_level1"));
+    }
```
Comment (alexjo2144) [inline plugin/trino-hive/src/test/java/io/trino/plugin/hive/parquet/TestParquetPageSourceFactory.java:57]: Nit pick on the formatting, Typically if method/constructor arguments need to be split over lines each one gets their own line. 

```suggestion
    @Test
    public void testGetNestedMixedRepetitionColumnType()
    {
        RowType rowType = rowType(
                RowType.field(
                        "optional_level2",
                        rowType(RowType.field(
                                "required_level3",
                                IntegerType.INTEGER))));
        HiveColumnHandle columnHandle = new HiveColumnHandle(
                "optional_level1",
                0,
                HiveType.valueOf("struct<optional_level2:struct<required_level3:int>>"),
                rowType,
                Optional.of(
                        new HiveColumnProjectionInfo(
                                ImmutableList.of(1, 1),
                                ImmutableList.of("optional_level2", "required_level3"),
                                toHiveType(IntegerType.INTEGER),
                                IntegerType.INTEGER)),
                REGULAR,
                Optional.empty());
        MessageType fileSchema = new MessageType(
                "hive_schema",
                new GroupType(OPTIONAL, "optional_level1",
                        new GroupType(OPTIONAL, "optional_level2",
                                new PrimitiveType(REQUIRED, INT32, "required_level3"))));
        assertEquals(
                ParquetPageSourceFactory.getColumnType(columnHandle, fileSchema, true).get(),
                fileSchema.getType("optional_level1"));
    }
```

**Comment by alexjo2144** [comment]
Since we don't have a way to run Spark code in the test suite, only Spark SQL, we could generate a small parquet file and check it in to the repo, but a unit test like this is enough for me personally.

**Comment by lxqfy** [comment]
Hi, Any suggestions on how should I proceed?

**Comment by findepi** [comment]
Squashed and merged as 3dd434500843999f1608522a973a5db9b7524f41. Thanks!

**Pull request closed without merging**